### PR TITLE
Feature/jwt implementation

### PR DESCRIPTION
**Pull request by wkrzyzanowski**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,98 +1,232 @@
...
+            <!--<artifactId>spring-beans</artifactId>-->
+            <!--<version>LATEST</version>-->
+            <!--</dependency>-->
+            <!--<dependency>-->
+            <!--<groupId>org.springframework.data</groupId>-->
+            <!--<artifactId>spring-data-jpa</artifactId>-->
+            <!--<version>LATEST</version>-->
+            <!--</dependency>-->
+            <!--</dependencies>-->
+            <!--</plugin>-->
```
Comment (lukaszfranczuk) [inline pom.xml:226]: please remove commented code

**Review by miles87 — APPROVED** [review]

File: src/main/java/pl/betse/beontime/bo/DepartmentDTO.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+@ToString
+@AllArgsConstructor
+@NoArgsConstructor
+public class DepartmentDTO {
+
+    private int id;
+
+    private String name;
+
+    private List<UserDTO> users;
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/bo/DepartmentDTO.java:20]: It should be called ``DepartmentBo`` and contains only business relevant fields - ``guid`` and ``name``

File: src/main/java/pl/betse/beontime/bo/DepartmentDTO.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+import java.util.Objects;
+
+@Getter
+@Setter
+@Builder
+@ToString
+@AllArgsConstructor
+@NoArgsConstructor
+public class DepartmentDTO {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/bo/DepartmentDTO.java:14]: please use ``@Data`` annotation

File: src/main/java/pl/betse/beontime/bo/RoleDTO.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+@ToString
+@AllArgsConstructor
+@NoArgsConstructor
+public class RoleDTO {
+
+    private int id;
+
+    private String role;
+
+    private List<UserDTO> users;
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/bo/RoleDTO.java:21]: we should not have ``RoleBo`` class - roles are not business objects

File: src/main/java/pl/betse/beontime/bo/UserCredentials.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package pl.betse.beontime.bo;
+
+import lombok.Data;
+import pl.betse.beontime.model.validation.CredentialsValidation;
+
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+
+@Data
+public class UserCredentials {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/bo/UserCredentials.java:10]: why do we need such class?

File: src/main/java/pl/betse/beontime/bo/UserDTO.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.List;
+import java.util.Objects;
+
+@Getter
+@Setter
+@Builder
+@ToString
+@AllArgsConstructor
+@NoArgsConstructor
+public class UserDTO {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/bo/UserDTO.java:19]: should be called ``UserBo`` and contains only business relevant fields

File: src/main/java/pl/betse/beontime/bo/UserDTO.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import javax.validation.constraints.NotNull;
+import java.util.List;
+import java.util.Objects;
+
+@Getter
+@Setter
+@Builder
+@ToString
+@AllArgsConstructor
+@NoArgsConstructor
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/bo/UserDTO.java:18]: please use ``@Data`` annotation

File: src/main/java/pl/betse/beontime/controller/DepartmentController.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    private UsersService usersService;
+    private DepartmentService departmentService;
+
+    public DepartmentController(UsersService usersService, DepartmentService departmentService) {
+        this.usersService = usersService;
+        this.departmentService = departmentService;
+    }
+
+    @GetMapping
+    public List<DepartmentDTO> getDepartmentWithUsers() {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/DepartmentController.java:32]: ``DepartmentsController`` should return only list of departments - nothing else

File: src/main/java/pl/betse/beontime/controller/ErrorHandlerController.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+import pl.betse.beontime.model.exception.*;
+import pl.betse.beontime.utils.CustomResponseMessage;
+
+import java.util.ArrayList;
+
+@RestControllerAdvice
+public class ErrorHandlerController extends ResponseEntityExceptionHandler {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/ErrorHandlerController.java:20]: Please use logger when exception handling

File: src/main/java/pl/betse/beontime/controller/RoleController.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import pl.betse.beontime.service.RoleService;
+import pl.betse.beontime.utils.UserDTOListBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@RestController
+@RequestMapping("/roles")
+@CrossOrigin("*")
+public class RoleController {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/RoleController.java:19]: we shouldn't have endpoint for roles

File: src/main/java/pl/betse/beontime/controller/TestController.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+import pl.betse.beontime.bo.UserDTO;
+import pl.betse.beontime.model.exception.UserNotFoundException;
+import pl.betse.beontime.mapper.UserModelMapper;
+import pl.betse.beontime.service.UsersService;
+
+@RestController
+public class TestController {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/TestController.java:14]: Please remove test controller - it is not business relevant

File: src/main/java/pl/betse/beontime/controller/UserController.java [hunk 12]
```diff
@@ -1,13 +1,192 @@
...
+import pl.betse.beontime.utils.GUIDGenerator;
+import pl.betse.beontime.utils.UpperAndLoweCaseCorrector;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping("/")
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/UserController.java:29]: please add ``/users`` request mapping and change gateway redirect

File: src/main/java/pl/betse/beontime/controller/UserController.java [hunk 13]
```diff
@@ -1,13 +1,192 @@
...
+
+        if (!usersService.existsByGUID(userGUID)) {
+            throw new UserNotFoundException();
+        }
+
+        return UserModelMapper.fromUserEntityToUserDTO(usersService.findByGUID(userGUID));
+    }
+
+    @GetMapping(path = "/email")
+    public @ResponseBody
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/UserController.java:69]: please remove this method

File: src/main/java/pl/betse/beontime/controller/UserController.java [hunk 14]
```diff
@@ -1,13 +1,192 @@
...
+
+        if (!EnumUtils.isValidEnum(DepartmentEnum.class, userDTO.getDepartment().toUpperCase())) {
+            throw new DepartmentNotFoundException();
+        }
+
+        Set<RoleEntity> newRoleEntities = new HashSet<>();
+        if (userDTO.getRoles() != null) {
+            if (validateUserRoles(userDTO, newRoleEntities))
+                throw new RoleNotFoundException();
+        }
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/UserController.java:94]: all this validations should happen in service - service should be responsible for business logic

File: src/main/java/pl/betse/beontime/controller/UserController.java [hunk 15]
```diff
@@ -1,13 +1,192 @@
...
+
+        UserEntity newUserEntity = UserModelMapper.fromUserDtoToUserEntity(userDTO);
+        // SET PASSWORD FOR EVERY NEW USER (HARDCODED FOR NOW!)
+        newUserEntity.setPassword(passwordEncoder.encode("qwe123!"));
+        newUserEntity.setDepartmentEntity(departmentService.findByName(userDTO.getDepartment()));
+        newUserEntity.setRoles(newRoleEntities);
+
+        usersService.save(newUserEntity);
+
+        return UserModelMapper.fromUserEntityToUserDTO(newUserEntity);
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/controller/UserController.java:114]: please don't use static methods

File: src/main/java/pl/betse/beontime/entity/RoleEntity.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "ROLE_ID")
+    private int id;
+
+    @Column(nullable = false, unique = true)
+    private String role;
+
+    @ManyToMany(mappedBy = "roles")
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/entity/RoleEntity.java:26]: please add ``fetch = FetchType.LAZY`` to that mapping - we should fetch all users for a role

File: src/main/java/pl/betse/beontime/entity/UserEntity.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table(name = "USER")
+public class UserEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "user_ID")
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/entity/UserEntity.java:24]: all columns name UPPER_CASE

File: src/main/java/pl/betse/beontime/model/enums/DepartmentEnum.java [hunk 18]
```diff
@@ -0,0 +1,7 @@
+package pl.betse.beontime.model.enums;
+
+public enum DepartmentEnum {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/model/enums/DepartmentEnum.java:3]: please remove DepartmetnEnum - we should not use enums

File: src/main/java/pl/betse/beontime/model/enums/RoleEnum.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package pl.betse.beontime.model.enums;
+
+public enum RoleEnum {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/model/enums/RoleEnum.java:3]: please remove RoleEnum - we should not use enums

File: src/main/java/pl/betse/beontime/model/exception/DepartmentNotFoundException.java [hunk 20]
```diff
@@ -0,0 +1,7 @@
+package pl.betse.beontime.model.exception;
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+public class DepartmentNotFoundException extends NullPointerException {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/model/exception/DepartmentNotFoundException.java:6]: custom exceptions should extends ``RuntimeException``

File: src/main/java/pl/betse/beontime/model/validation/LoginUserValidation.java [hunk 21]
```diff
@@ -0,0 +1,4 @@
+package pl.betse.beontime.model.validation;
+
+public interface LoginUserValidation {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/model/validation/LoginUserValidation.java:3]: Do we need that empty interfaces?

File: src/main/java/pl/betse/beontime/service/DepartmentServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import pl.betse.beontime.entity.DepartmentEntity;
+import pl.betse.beontime.repository.DepartamentRepository;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class DepartmentServiceImpl implements DepartmentService {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/service/DepartmentServiceImpl.java:12]: please use just classes - we don't interfaces with annotation configuration

File: src/main/java/pl/betse/beontime/utils/CustomResponseMessage.java [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+
+import lombok.*;
+import org.springframework.http.HttpStatus;
+
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class CustomResponseMessage {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/utils/CustomResponseMessage.java:12]: we don't want to have any custom response, especially with fields that are already in HTTP Message

File: src/main/java/pl/betse/beontime/utils/UpperAndLoweCaseCorrector.java [hunk 24]
```diff
@@ -0,0 +1,8 @@
+package pl.betse.beontime.utils;
+
+public class UpperAndLoweCaseCorrector {
+
+    public static String fix(String toCorrect) {
```
Comment (lukaszfranczuk) [inline src/main/java/pl/betse/beontime/utils/UpperAndLoweCaseCorrector.java:5]: ``fix`` is not very descriptive name

File: src/main/resources/application.yaml [hunk 25]
```diff
@@ -0,0 +1,34 @@
+server:
+  port: ${SERVER_PORT:8080}
```
Comment (lukaszfranczuk) [inline src/main/resources/application.yaml:2]: please don't commit ``application.yml`` - it is only for local env

File: src/test/java/pl/betse/beontime/LoginControllerTest.java [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+    @BeforeEach
+    void setup() {
+        TEST_USER = new UserDTO("CustomGUID", "test@test.pl", "TestName", "TestLastName", false, "BANKING", new ArrayList<>());
+        TEST_USER_ENTITY = new UserEntity(1, "CustomGUID", TEST_USER.getEmail(), TEST_USER.getFirstName(), TEST_USER.getLastName(), "PASSWORD", TEST_USER.isActive(), new DepartmentEntity(1, "BANKING"), new HashSet<>());
+        loginController = new LoginController(userService, passwordEncoder);
+    }
+
+
+    @Test
+    void checkExsistsByEmailLogin() {
```
Comment (lukaszfranczuk) [inline src/test/java/pl/betse/beontime/LoginControllerTest.java:40]: test should start with ``should`` and describe what tested method should return or do

File: src/main/java/pl/betse/beontime/bo/UserCredentials.java (hunk 5 above)
Comment (wkrzyzanowski) [reply src/main/java/pl/betse/beontime/bo/UserCredentials.java:10]: Now, when authentication-service works, we don't need it. So it should be deleted in refactoring process.

File: src/main/java/pl/betse/beontime/controller/ErrorHandlerController.java (hunk 9 above)
Comment (wkrzyzanowski) [reply src/main/java/pl/betse/beontime/controller/ErrorHandlerController.java:20]: Lombok annotation `@Slf4j` will be a good way?

File: src/main/java/pl/betse/beontime/model/validation/LoginUserValidation.java (hunk 21 above)
Comment (wkrzyzanowski) [reply src/main/java/pl/betse/beontime/model/validation/LoginUserValidation.java:3]: If we're using a `@Validated(ExampleValidation.class)` in controller methods and in BO fileds marked by validation annotations ex. `@NotNull(groups={ExampleValidation.class})` we need this empty interefaces - I don't know easier way to validate incoming JSON objects in various scenarios.

**Pull request merged**